File: packages/runtime-ts/tests/bit-reader.test.ts
import { describe, it, expect } from 'vitest';
import { BitReader } from '../src/bit-reader.js';
import { BitWriter } from '../src/bit-writer.js';

describe('BitReader', () => {
  describe('readBits', () => {
    it('reads a single true bit', () => {
      const r = new BitReader(new Uint8Array([0x01]));
      expect(r.readBool()).toBe(true);
    });

    it('reads a single false bit', () => {
      const r = new BitReader(new Uint8Array([0x00]));
      expect(r.readBool()).toBe(false);
    });

    it('reads sub-byte fields LSB-first', () => {
      // 0x9D = 10011101 -> LSB-first: bits[0..3] = 101 = 5, bits[3..8] = 10011 = 19
      const r = new BitReader(new Uint8Array([0x9d]));
      expect(r.readBits(3)).toBe(5);
      expect(r.readBits(5)).toBe(19);
    });

    it('reads across byte boundary', () => {
      const r = new BitReader(new Uint8Array([0x9d, 0x2a]));
      expect(r.readBits(3)).toBe(5);
      expect(r.readBits(5)).toBe(19);
      expect(r.readBits(6)).toBe(42);
    });
  });

  describe('flushToByteBoundary', () => {
    it('skips remaining bits in current byte', () => {
      const w = new BitWriter();
      w.writeBits(0b101, 3);
      w.flushToByteBoundary();
      w.writeU8(0xab);
      const buf = w.finish();

      const r = new BitReader(buf);
      expect(r.readBits(3)).toBe(0b101);
      r.flushToByteBoundary();
      expect(r.readU8()).toBe(0xab);
    });
  });

  describe('multi-byte reads', () => {
    it('readU8', () => {
      const r = new BitReader(new Uint8Array([0xff]));
      expect(r.readU8()).toBe(255);
    });

    it('readU16 little-endian', () => {
      const r = new BitReader(new Uint8Array([0x02, 0x01]));
      expect(r.readU16()).toBe(258);
    });

    it('readU32 little-endian', () => {
      const r = new BitReader(new Uint8Array([0x78, 0x56, 0x34, 0x12]));
      expect(r.readU32()).toBe(0x12345678);
    });

    it('readU64 little-endian', () => {
      const r = new BitReader(
        new Uint8Array([0x08, 0x07, 0x06, 0x05, 0x04, 0x03, 0x02, 0x01]),
      );
      expect(r.readU64()).toBe(BigInt('0x0102030405060708'));
    });

    it('readI8 negative', () => {
      const r = new BitReader(new Uint8Array([0xff]));
      expect(r.readI8()).toBe(-1);
    });

    it('readI16 negative', () => {
      const r = new BitReader(new Uint8Array([0xff, 0xff]));
      expect(r.readI16()).toBe(-1);
    });

    it('readI32 negative', () => {
      const r = new BitReader(new Uint8Array([0xff, 0xff, 0xff, 0xff]));
      expect(r.readI32()).toBe(-1);
    });

    it('readI64 negative', () => {
      const r = new BitReader(
        new Uint8Array([0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff]),
      );
      expect(r.readI64()).toBe(BigInt(-1));
    });
  });

  describe('floating point', () => {
    it('readF32 NaN canonical', () => {
      const r = new BitReader(new Uint8Array([0x00, 0x00, 0xc0, 0x7f]));
      const v = r.readF32();
      expect(Number.isNaN(v)).toBe(true);
    });

    it('readF64 negative zero', () => {
      const r = new BitReader(
        new Uint8Array([0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x80]),
      );
      const v = r.readF64();
      expect(v).toBe(-0.0);
      expect(Object.is(v, -0)).toBe(true);
    });

    it('readF32 pi round-trip', () => {
      const w = new BitWriter();
      w.writeF32(Math.fround(Math.PI));
      const buf = w.finish();
      const r = new BitReader(buf);
      expect(r.readF32()).toBe(Math.fround(Math.PI));
    });

    it('readF64 NaN round-trip is canonical', () => {
      const w = new BitWriter();
      w.writeF64(NaN);
      const buf = w.finish();
      const r = new BitReader(buf);
      const v = r.readF64();
      expect(Number.isNaN(v)).toBe(true);
    });
  });

  describe('LEB128', () => {
    it('reads 0', () => {
      const r = new BitReader(new Uint8Array([0x00]));
      expect(r.readLeb128()).toBe(0);
    });

    it('reads 300', () => {
      const r = new BitReader(new Uint8Array([0xac, 0x02]));
      expect(r.readLeb128()).toBe(300);
    });

    it('round-trip 128', () => {
      const w = new BitWriter();
      w.writeLeb128(128);
      const r = new BitReader(w.finish());
      expect(r.readLeb128()).toBe(128);
    });
  });

  describe('string and bytes', () => {
    it('readString "hello"', () => {
      const r = new BitReader(
        new Uint8Array([0x05, 0x68, 0x65, 0x6c, 0x6c, 0x6f]),
      );
      expect(r.readString()).toBe('hello');
    });

    it('readString empty', () => {
      const r = new BitReader(new Uint8Array([0x00]));
      expect(r.readString()).toBe('');
    });

    it('readBytes', () => {
      const r = new BitReader(new Uint8Array([0x02, 0xde, 0xad]));
      expect(r.readBytes()).toEqual(new Uint8Array([0xde, 0xad]));
    });

    it('readRawBytes', () => {
      const r = new BitReader(new Uint8Array([0xca, 0xfe]));
      expect(r.readRawBytes(2)).toEqual(new Uint8Array([0xca, 0xfe]));
    });

    it('string round-trip', () => {
      const w = new BitWriter();
      w.writeString('hello');
      const r = new BitReader(w.finish());
      expect(r.readString()).toBe('hello');
    });
  });

  describe('error handling', () => {
    it('throws on unexpected EOF for readU8', () => {
      const r = new BitReader(new Uint8Array([]));
      expect(() => r.readU8()).toThrow('Unexpected end of data');
    });

    it('throws on unexpected EOF for readBits', () => {
      const r = new BitReader(new Uint8Array([]));
      expect(() => r.readBits(1)).toThrow('Unexpected end of data');
    });

    it('throws on unexpected EOF for readU16', () => {
      const r = new BitReader(new Uint8Array([0x01]));
      expect(() => r.readU16()).toThrow('Unexpected end of data');
    });
  });

  describe('remaining', () => {
    it('reports remaining bytes', () => {
      const r = new BitReader(new Uint8Array([0x01, 0x02, 0x03]));
      expect(r.remaining()).toBe(3);
      r.readU8();
      expect(r.remaining()).toBe(2);
    });
  });

  describe('readLeb12864', () => {
    it('round-trips 0n', () => {
      const w = new BitWriter();
      w.writeLeb12864(0n);
      const r = new BitReader(w.finish());
      expect(r.readLeb12864()).toBe(0n);
    });

    it('round-trips 300n', () => {
      const w = new BitWriter();
      w.writeLeb12864(300n);
      const r = new BitReader(w.finish());
      expect(r.readLeb12864()).toBe(300n);
    });

    it('round-trips large value', () => {
      const w = new BitWriter();
      const big = 1234567890123456789n;
      w.writeLeb12864(big);
      const r = new BitReader(w.finish());
      expect(r.readLeb12864()).toBe(big);
    });
  });

  describe('readZigZag', () => {
    it('decodes 0', () => {
      const w = new BitWriter();
      w.writeZigZag(0, 32);
      const r = new BitReader(w.finish());
      expect(r.readZigZag(32)).toBe(0);
    });

    it('decodes -1', () => {
      const w = new BitWriter();
      w.writeZigZag(-1, 32);
      const r = new BitReader(w.finish());
      expect(r.readZigZag(32)).toBe(-1);
    });

    it('decodes 1', () => {
      const w = new BitWriter();
      w.writeZigZag(1, 32);
      const r = new BitReader(w.finish());
      expect(r.readZigZag(32)).toBe(1);
    });

    it('round-trips -42', () => {
      const w = new BitWriter();
      w.writeZigZag(-42, 32);
      const r = new BitReader(w.finish());
      expect(r.readZigZag(32)).toBe(-42);
    });
  });

  describe('readZigZag64', () => {
    it('round-trips 0n', () => {
      const w = new BitWriter();
      w.writeZigZag64(0n);
      const r = new BitReader(w.finish());
      expect(r.readZigZag64()).toBe(0n);
    });

    it('round-trips -1n', () => {
      const w = new BitWriter();
      w.writeZigZag64(-1n);
      const r = new BitReader(w.finish());
      expect(r.readZigZag64()).toBe(-1n);
    });

    it('round-trips large negative', () => {
      const w = new BitWriter();
      w.writeZigZag64(-123456789n);
      const r = new BitReader(w.finish());
      expect(r.readZigZag64()).toBe(-123456789n);
    });
  });

  describe('recursion depth', () => {
    it('allows up to 64 levels', () => {
      const r = new BitReader(new Uint8Array([]));
      for (let i = 0; i < 64; i++) {
        r.enterNested();
      }
    });

    it('throws at 65 levels', () => {
      const r = new BitReader(new Uint8Array([]));
      for (let i = 0; i < 64; i++) {
        r.enterNested();
      }
      expect(() => r.enterNested()).toThrow('Recursion limit exceeded');
    });

    it('leave allows re-entry', () => {
      const r = new BitReader(new Uint8Array([]));
      for (let i = 0; i < 64; i++) {
        r.enterNested();
      }
      r.leaveNested();
      r.enterNested(); // Should not throw
    });
  });

  describe('round-trip with BitWriter', () => {
    it('round-trips sub-byte fields', () => {
      const w = new BitWriter();
      w.writeBits(5, 3);
      w.writeBits(19, 5);
      w.writeBits(42, 6);
      const buf = w.finish();
      const r = new BitReader(buf);
      expect(r.readBits(3)).toBe(5);
      expect(r.readBits(5)).toBe(19);
      expect(r.readBits(6)).toBe(42);
    });

    it('round-trips u16', () => {
      const w = new BitWriter();
      w.writeU16(0x1234);
      const r = new BitReader(w.finish());
      expect(r.readU16()).toBe(0x1234);
    });

    it('round-trips i32 negative', () => {
      const w = new BitWriter();
      w.writeI32(-42);
      const r = new BitReader(w.finish());
      expect(r.readI32()).toBe(-42);
    });

    it('round-trips mixed types', () => {
      const w = new BitWriter();
      w.writeBool(true);
      w.writeU16(42);
      w.writeString('test');
      const buf = w.finish();
      const r = new BitReader(buf);
      expect(r.readBool()).toBe(true);
      r.flushToByteBoundary();
      expect(r.readU16()).toBe(42);
      expect(r.readString()).toBe('test');
    });
  });

  describe('readRemaining', () => {
    it('reads remaining bytes after partial decode', () => {
      const r = new BitReader(new Uint8Array([0x2a, 0x00, 0x00, 0x00, 0x63, 0x00]));
      r.readU32();
      const remaining = r.readRemaining();
      expect(remaining.length).toBe(2);
      expect(remaining[0]).toBe(0x63);
      expect(remaining[1]).toBe(0x00);
    });

    it('returns empty when fully consumed', () => {
      const r = new BitReader(new Uint8Array([0x2a, 0x00, 0x00, 0x00]));
      r.readU32();
      const remaining = r.readRemaining();
      expect(remaining.length).toBe(0);
    });

    it('reads all bytes from start', () => {
      const r = new BitReader(new Uint8Array([0x01, 0x02, 0x03]));
      const remaining = r.readRemaining();
      expect(remaining).toEqual(new Uint8Array([0x01, 0x02, 0x03]));
    });
  });

  describe('trailing bytes tolerance', () => {
    it('does not reject trailing bytes', () => {
      // Simulate v2-encoded message read by v1 decoder
      const data = new Uint8Array([0x2a, 0x00, 0x00, 0x00, 0x63, 0x00]);
      const r = new BitReader(data);
      expect(r.readU32()).toBe(42);
      r.flushToByteBoundary();
      // Remaining bytes must not cause error
      expect(r.remaining()).toBe(2);
    });
  });
});

File: packages/runtime-go/bitreader_test.go
package vexil

import (
	"errors"
	"math"
	"testing"
)

func TestReadSingleBit(t *testing.T) {
	r := NewBitReader([]byte{0x01})
	v, err := r.ReadBool()
	assertNoError(t, err)
	if !v {
		t.Fatal("expected true")
	}
}

func TestRoundTripSubByte(t *testing.T) {
	w := NewBitWriter()
	w.WriteBits(5, 3)
	w.WriteBits(19, 5)
	w.WriteBits(42, 6)
	buf := w.Finish()
	r := NewBitReader(buf)

	v, err := r.ReadBits(3)
	assertNoError(t, err)
	assertUint64(t, v, 5)

	v, err = r.ReadBits(5)
	assertNoError(t, err)
	assertUint64(t, v, 19)

	v, err = r.ReadBits(6)
	assertNoError(t, err)
	assertUint64(t, v, 42)
}

func TestRoundTripU8(t *testing.T) {
	w := NewBitWriter()
	w.WriteU8(0xAB)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadU8()
	assertNoError(t, err)
	if v != 0xAB {
		t.Fatalf("got %02X, want AB", v)
	}
}

func TestRoundTripU16(t *testing.T) {
	w := NewBitWriter()
	w.WriteU16(0x1234)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadU16()
	assertNoError(t, err)
	if v != 0x1234 {
		t.Fatalf("got %04X, want 1234", v)
	}
}

func TestRoundTripU32(t *testing.T) {
	w := NewBitWriter()
	w.WriteU32(0x12345678)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadU32()
	assertNoError(t, err)
	if v != 0x12345678 {
		t.Fatalf("got %08X, want 12345678", v)
	}
}

func TestRoundTripU64(t *testing.T) {
	w := NewBitWriter()
	w.WriteU64(0x0102030405060708)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadU64()
	assertNoError(t, err)
	if v != 0x0102030405060708 {
		t.Fatalf("got %016X, want 0102030405060708", v)
	}
}

func TestRoundTripI32Neg(t *testing.T) {
	w := NewBitWriter()
	w.WriteI32(-42)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadI32()
	assertNoError(t, err)
	if v != -42 {
		t.Fatalf("got %d, want -42", v)
	}
}

func TestRoundTripI64(t *testing.T) {
	w := NewBitWriter()
	w.WriteI64(-1000)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadI64()
	assertNoError(t, err)
	if v != -1000 {
		t.Fatalf("got %d, want -1000", v)
	}
}

func TestRoundTripF32(t *testing.T) {
	w := NewBitWriter()
	w.WriteF32(math.Pi)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadF32()
	assertNoError(t, err)
	if v != float32(math.Pi) {
		t.Fatalf("got %v, want %v", v, float32(math.Pi))
	}
}

func TestRoundTripF64Nan(t *testing.T) {
	w := NewBitWriter()
	w.WriteF64(math.NaN())
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadF64()
	assertNoError(t, err)
	if !math.IsNaN(v) {
		t.Fatal("expected NaN")
	}
	if math.Float64bits(v) != 0x7FF8000000000000 {
		t.Fatalf("got %016X, want canonical NaN", math.Float64bits(v))
	}
}

func TestRoundTripString(t *testing.T) {
	w := NewBitWriter()
	w.WriteString("hello")
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadString()
	assertNoError(t, err)
	if v != "hello" {
		t.Fatalf("got %q, want %q", v, "hello")
	}
}

func TestRoundTripStringEmpty(t *testing.T) {
	w := NewBitWriter()
	w.WriteString("")
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadString()
	assertNoError(t, err)
	if v != "" {
		t.Fatalf("got %q, want empty", v)
	}
}

func TestRoundTripLeb128(t *testing.T) {
	w := NewBitWriter()
	w.WriteLeb128(300)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadLeb128(4)
	assertNoError(t, err)
	assertUint64(t, v, 300)
}

func TestRoundTripZigZag(t *testing.T) {
	w := NewBitWriter()
	w.WriteZigZag(-42, 64)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadZigZag(64, 10)
	assertNoError(t, err)
	if v != -42 {
		t.Fatalf("got %d, want -42", v)
	}
}

func TestRoundTripBytes(t *testing.T) {
	w := NewBitWriter()
	w.WriteBytes([]byte{0xDE, 0xAD})
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadBytes()
	assertNoError(t, err)
	assertEqual(t, v, []byte{0xDE, 0xAD})
}

func TestRoundTripRawBytes(t *testing.T) {
	w := NewBitWriter()
	w.WriteRawBytes([]byte{0xCA, 0xFE})
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadRawBytes(2)
	assertNoError(t, err)
	assertEqual(t, v, []byte{0xCA, 0xFE})
}

func TestReadUnexpectedEOF(t *testing.T) {
	r := NewBitReader([]byte{})
	_, err := r.ReadU8()
	if !errors.Is(err, ErrUnexpectedEOF) {
		t.Fatalf("expected ErrUnexpectedEOF, got %v", err)
	}
}

func TestReadInvalidUTF8(t *testing.T) {
	w := NewBitWriter()
	w.WriteLeb128(2)
	w.WriteRawBytes([]byte{0xFF, 0xFE})
	buf := w.Finish()
	r := NewBitReader(buf)
	_, err := r.ReadString()
	if !errors.Is(err, ErrInvalidUTF8) {
		t.Fatalf("expected ErrInvalidUTF8, got %v", err)
	}
}

func TestReaderRecursionDepthLimit(t *testing.T) {
	r := NewBitReader([]byte{})
	for i := 0; i < 64; i++ {
		if err := r.EnterRecursive(); err != nil {
			t.Fatalf("depth %d: %v", i+1, err)
		}
	}
	if err := r.EnterRecursive(); err == nil {
		t.Fatal("expected error at depth 65")
	}
}

func TestReaderRecursionDepthLeave(t *testing.T) {
	r := NewBitReader([]byte{})
	for i := 0; i < 64; i++ {
		assertNoError(t, r.EnterRecursive())
	}
	r.LeaveRecursive()
	assertNoError(t, r.EnterRecursive())
}

func TestTrailingBytesNotRejected(t *testing.T) {
	data := []byte{0x2a, 0x00, 0x00, 0x00, 0x63, 0x00}
	r := NewBitReader(data)
	v, err := r.ReadU32()
	assertNoError(t, err)
	if v != 42 {
		t.Fatalf("got %d, want 42", v)
	}
	r.FlushToByteBoundary()
	// Remaining bytes must not cause error
}

func TestReadRemainingAfterPartialDecode(t *testing.T) {
	data := []byte{0x2a, 0x00, 0x00, 0x00, 0x63, 0x00}
	r := NewBitReader(data)
	_, _ = r.ReadU32()
	remaining := r.ReadRemaining()
	assertEqual(t, remaining, []byte{0x63, 0x00})
}

func TestReadRemainingWhenFullyConsumed(t *testing.T) {
	data := []byte{0x2a, 0x00, 0x00, 0x00}
	r := NewBitReader(data)
	_, _ = r.ReadU32()
	remaining := r.ReadRemaining()
	if len(remaining) != 0 {
		t.Fatalf("expected empty, got %X", remaining)
	}
}

func TestReadRemainingFromStart(t *testing.T) {
	data := []byte{0x01, 0x02, 0x03}
	r := NewBitReader(data)
	remaining := r.ReadRemaining()
	assertEqual(t, remaining, []byte{0x01, 0x02, 0x03})
}

func TestReaderFlushToByteBoundary(t *testing.T) {
	w := NewBitWriter()
	w.WriteBits(0b101, 3)
	w.FlushToByteBoundary()
	w.WriteU8(0xAB)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadBits(3)
	assertNoError(t, err)
	assertUint64(t, v, 0b101)
	r.FlushToByteBoundary()
	u, err := r.ReadU8()
	assertNoError(t, err)
	if u != 0xAB {
		t.Fatalf("got %02X, want AB", u)
	}
}

func TestReadBitsFromMultipleBytes(t *testing.T) {
	// Write 3+5+6 = 14 bits across 2 bytes
	w := NewBitWriter()
	w.WriteBits(7, 3)  // 111
	w.WriteBits(31, 5) // 11111
	w.WriteBits(63, 6) // 111111
	buf := w.Finish()

	r := NewBitReader(buf)
	a, err := r.ReadBits(3)
	assertNoError(t, err)
	assertUint64(t, a, 7)

	b, err := r.ReadBits(5)
	assertNoError(t, err)
	assertUint64(t, b, 31)

	c, err := r.ReadBits(6)
	assertNoError(t, err)
	assertUint64(t, c, 63)
}

func TestReadI8(t *testing.T) {
	w := NewBitWriter()
	w.WriteI8(-1)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadI8()
	assertNoError(t, err)
	if v != -1 {
		t.Fatalf("got %d, want -1", v)
	}
}

func TestReadI16(t *testing.T) {
	w := NewBitWriter()
	w.WriteI16(-1000)
	buf := w.Finish()
	r := NewBitReader(buf)
	v, err := r.ReadI16()
	assertNoError(t, err)
	if v != -1000 {
		t.Fatalf("got %d, want -1000", v)
	}
}

// helper functions

func assertNoError(t *testing.T, err error) {
	t.Helper()
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func assertUint64(t *testing.T, got, want uint64) {
	t.Helper()
	if got != want {
		t.Fatalf("got %d, want %d", got, want)
	}
}
